### ISSUE TITLE
Can not connect with Revolut bank

### ISSUE THREAD

**Issue by LukaFinzgar** [plaid/plaid-link-android#92]
Greetings. 

We have trouble connecting Revolut using Plaid SDK for android (version 2.1.0). 
We start the process using standard PLaid.openLink (….) call.
User successfully authenticates using Revolut app and is then asked to return to our app, but after confirming that nothing happens.
Looking at logs, i can see something like that:
"2020-08-21 10:45:25.670 2073-6897/? I/ActivityTaskManager: START u0 {act=android.intent.action.VIEW cat=[android.intent.category.DEFAULT,android.intent.category.BROWSABLE] dat=plaid://redirect/?oauth_state_id=d114313d-e517-41df-bbd6-50b5f69f1324 pkg=com.thirdframestudios.android.expensoor cmp=com.thirdframestudios.android.expensoor/com.plaid.link.internal.LinkRedirectActivity (has extras)} from uid 10231"

https://plaid.com/docs/#oauth says that android app does not need any changes in order to support these OAuth authentication flows, but at the end of this chapter ("App links") it says that android deep links are required. 
Do we need to add some deep links (with scheme "plaid") to our android app to support OAuth authentication flow?

**Comment by slauzinho** [comment]
I have the same issue.

Testing with a sandbox account after `Grant access` it loads a page in chrome but it doesn't redirect back to the app.

I'm wondering if we need to configure App Links

**Comment by zsweigart** [comment]
Hi @LukaFinzgar and @slauzinho 
Inside the Android sdk we build a redirect url using the [intent format](https://developer.chrome.com/multidevice/android/intents) which includes your app package name so that it cannot be intercepted by other apps.  The sdk also includes an intent filter for urls that match `plaid://redirect`.  You do not need to configure app links; however, you do need to [whitelist your package name](https://plaid.com/docs/link/android/#register-app-id)  Could you please check that your package name is whitelisted - from your logs  it looks like it would be `com.thirdframestudios.android.expensoor`.

**Comment by slauzinho** [comment]
Hi @zsweigart I can confirm that I have my package name whitelisted. I even tested cloning this repo and add `com.plaid.linksample` to the whitelist but I still have the same problem.

Looking at the logcat I get this input after being redirected to chrome:


```txt 
2020-08-25 20:54:32.200 978-978/? E/cr_InstantAppsHandlerI: Exception when calling getInstantAppLaunchData
    zt: 19500: Not an instant app
        at hw.a(chromium-TrichromeChromeGoogle.aab-stable-414712583:3)
        at Ww.a(chromium-TrichromeChromeGoogle.aab-stable-414712583:5)
        at com.google.android.gms.common.api.internal.BasePendingResult.k(chromium-TrichromeChromeGoogle.aab-stable-414712583:16)
        at com.google.android.gms.common.api.internal.BasePendingResult.j(chromium-TrichromeChromeGoogle.aab-stable-414712583:5)
        at SB.a1(chromium-TrichromeChromeGoogle.aab-stable-414712583:28)
        at MD.onTransact(chromium-TrichromeChromeGoogle.aab-stable-414712583:3)
        at android.os.Binder.execTransactInternal(Binder.java:1021)
        at android.os.Binder.execTransact(Binder.java:994)
```


Don't know if it's related.

**Comment by LukaFinzgar** [comment]
@zsweigart yes, we have com.thirdframestudios.android.expensoor whitelisted.
<img width="527" alt="Screenshot 2020-08-26 at 10 21 20" src="https://user-images.githubusercontent.com/45970522/91279647-0a243380-e786-11ea-9c12-acb301c1e49e.png">

**Comment by zsweigart** [comment]
I couldn't reproduce, a few questions though:

1. Are you using token configurations or public key configurations?
1. What device are you testing on? Is it reproducible on all devices?
1. Does your app have an instant app associated with it?

**Comment by slauzinho** [comment]
1- I'm using the token configuration
2- Tested on 2 physical phones (Samsung S10+ android 10 and Huawei P10 Lite android 8). Doesn't work on any device
3- Don't know what you mean by that. I'm trying on my app but also on the sample app of this repo.

-Here is what I tried:
`git clone https://github.com/plaid/plaid-link-android.git`

- Whitelist `com.plaid.linksample` inside plaid settings

- Generate link token by doing:
```
POST {
    "client_id": "xxxxx",
  "secret": "xxxx",
  "client_name": "My App",
  "user": { "client_user_id": "random_ui" },
  "products": ["auth"],
  "country_codes": ["GB"],
  "language": "en"
}
```

- Use the generated token directly on the app
- run `./gradlew installDebug`
- Click on link
- Choose random UK bank

I also made a video with the issue:
https://youtu.be/o0Nf30Sd_Z0

**Comment by LukaFinzgar** [comment]
1.) Using public key configuration
2.) Xiaomi Pocophone F1 (Android 9.0), Pixel 3 emulator (Android 10)
3.) We dont have instant app.

**Comment by zsweigart** [comment]
@slauzinho when you create your link token you will need to include the `android_package_name' parameter https://plaid.com/docs/#create-link-token

@LukaFinzgar can you locate your output apk (either /build/outputs/debug or  /build/outputs/release) and then use the apk analyzer (Build -> Analyze APK) to check that your AndroidManifest includes the Plaid activities
<img width="488" alt="Screen Shot 2020-08-27 at 1 11 49 PM" src="https://user-images.githubusercontent.com/1391816/91490280-d1648700-e877-11ea-8005-f2534cde5fbe.png">

**Comment by slauzinho** [comment]
@zsweigart It worked I don't know how I missed that. Thank you for your trouble 👍

**Comment by LukaFinzgar** [comment]
I analyzed apk and it contains  "com.plaid.link.internal.LinkRedirectActivity"  and ""com.plaid.link.internal.LinkActivity".

**Comment by LukaFinzgar** [comment]
Now we tried with plaid_token instead of public key configurations, still doesnt work.
`     ArrayList<PlaidProduct> products = new ArrayList<>();
        products.add(PlaidProduct.TRANSACTIONS);
        HashMap<String, String> extraParams = new HashMap<>();
        extraParams.put("institution", institution.getExt_id());
        Plaid.openLink(this,
                new LinkConfiguration.Builder()
                        .clientName(PlaidLinkParams.getPlaidClientName())
                        .products(products)
                        .environment(PlaidEnvironment.PRODUCTION)
                        .extraParams(extraParams)
                        .token(bankInstitution.getToken())
                        .build());`

In the past, when using webview without plaid sdk, we had to add params for oauth flow:
`.appendQueryParameter("oauthRedirectUri", ".ourAddress")
 .appendQueryParameter("oauthNonce", SyncUtils.generateRandomId())`

This is not needed for plaid SDK android afaik?

**Comment by LukaFinzgar** [comment]
Also we dont see any logs on Dashboard logs, eventhough server makes this:
`// 1. Grab the client_user_id by searching for the current user in your database
    User userFromDB = db.find(...);
    String clientUserId = userFromDB.id;
    LinkTokenCreateRequest.User user = new LinkTokenCreateRequest.User(clientUserId)`

**Comment by zsweigart** [comment]
For the token implementation you should use the `LinkTokenConfiguration` instead of `LinkConfiguration`.  `LinkTokenConfiguration` only requires the token and has fields for `logLevel` and `extraParams`.

To test on the client side you can generate a token for GB with institution Revolut and the transactions product using a curl command
```
curl -X POST https://sandbox.plaid.com/link/token/create -H 'Content-Type: application/json' -d '{ "client_id": "<YOUR_CLIENT_ID>", "secret": "<YOUR_SANDBOX_SECRET>", "user": { "client_user_id": "123456-unique"}, "client_name": "Expensoor", "products": ["transactions"], "country_codes":["GB"], "language":"en", "android_package_name":"com.thirdframestudios.android.expensoor", "institution_id":"ins_63"}
```

The token should look like
```
link-sandbox-0e4c1afd-323c-4025-b2dc-25b392b61328
```

Then to open Link
```
Plaid.openLink(this,
  new LinkTokenConfiguration.Builder()
    .token("<TOKEN_FROM_ABOVE")
    .build()
```

Also I would recommend that you upgrade 2.2.0, it is a non-breaking minor version bump so no additional code will be required.

**Comment by LukaFinzgar** [comment]
Now using 2.2.0 with LinkTokenConfiguration, Revolut still does not connect.  Tested in production env.

**Comment by zsweigart** [comment]
Hi @LukaFinzgar is this issue only reproduced for Revolut in production? 

Are you able to  get other banks to work successfully in sandbox and/or production?

**Comment by zsweigart** [comment]
Also @LukaFinzgar could you check the logs for `linkSessionId` so we can check any logs we have on our end

**Comment by LukaFinzgar** [comment]
8633144c-9e00-48e3-8654-c06561f65492

**Comment by zsweigart** [comment]
Hi @LukaFinzgar For that session id it does not look like oauth was opened so I'm not seeing any errors.  Would it be possible for you to provide us with a small sample app that reproduces the issue?

**Comment by LukaFinzgar** [comment]
Hi, you can try to connect Revolut in our beta Toshl app with Plaid sdk here (open side menu and go to "bank connections"): 
https://app.bitrise.io/artifact/48012941/p/9fde8c311f135d11aafa4d8def302640
username: luka@toshl.com
pass: test1234

**Comment by zsweigart** [comment]
Hi  Luka, I am able to see the error in the build you provided but I can't see any logs.  Could you provide the implementation code or a small separate app that  reproduces the issue easily? 

We especially need the code that you  are using to generate the link token and the code that you are using to open link.

**Comment by LukaFinzgar** [comment]
We generate link token on our backend, our android app just requests it (based on debugging, token looks good):
`Response<LinkTokenCreateResponse> response = getPlaidClient()
				.service()
				.linkTokenCreate(createLinkTokenRequest(institution))
				.execute();

	private LinkTokenCreateRequest createLinkTokenRequest(BankInstitution institution)
	{
		final String language = Stream.of("en", "fr", "es", "nl")
				.filter(l -> apiRequestInfo.getUserOfConcern().getLanguage().contains(l))
				.findAny()
				.orElse("en");
		return new LinkTokenCreateRequest(
				new LinkTokenCreateRequest.User(Long.toString(apiRequestInfo.getUser().getId())),
				"Toshl Finance",
				Collections.singletonList("transactions"),
				Arrays.asList("US", "GB", "CA", "FR", "ES", "NL"),
				language
		);
	}
`

Our android client calls link like this:
`HashMap<String, String> extraParams = new HashMap<>();
        extraParams.put("institution", institution.getExt_id());
        Plaid.openLink(this,
                new LinkTokenConfiguration.Builder()
                        .token(bankInstitution.getToken())
                        .extraParams(extraParams)
                        .build());`

**Comment by zsweigart** [comment]
This has been resolved in 3.0.1
